### PR TITLE
Enhancements to the coverage report

### PR DESCRIPTION
**Pull request by cicdguy**
Bugfixes for v2

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -138,14 +218,24 @@ runs:
...
+    - name: Fail if coverage worsened
+      run: |
+        if [[ "${{ inputs.coverage-reduction-failure }}" == "true" ]]
+        then {
+          if [[ ${{ env.pycobertura_status }} == "2" ]]
+          then {
+            echo "Code changes worsened the overall coverage."
+            exit 1
+          }
+          elif [[ ${{ env.pycobertura_status }} == "3" ]]
```
Comment (walkowif) [inline action.yml:230]: We have to decide if we want to fail in this case. Additionally, we can consider to fail when `status = 1`.
We can also introduce a new parameter to control this.
Quoting the [documentation](https://github.com/aconrad/pycobertura):
> Upon exit, the diff command may return various exit codes:
> 0: all changes are covered, no new uncovered statements have been introduced
> 1: some exception occurred (likely due to inappropriate usage or a bug in pycobertura)
> 2: the changes worsened the overall coverage
> 3: the changes introduced uncovered statements but the overall coverage is still better than before

**Comment by walkowif** [comment]
Example of comment where the code coverage is worse than on main:
https://github.com/insightsengineering/r.pkg.template/pull/100#issuecomment-1195199696

**Comment by dinakar29** [comment]
Looks good so far.
What happens to repositories which do not have the branch which stores the badges and coverage reports? Does the action create a new, empty orphan branch? If not, it should.

**Comment by walkowif** [comment]
Yeah, an orphan branch is created [here](https://github.com/insightsengineering/coverage-action/blob/179c7fd3ecaf44cbbfbe0aac188de22ec67b00d4/action.yml#L140). By empty do you mean that only the badge and the report should exist on this branch? Or can other files be there as well?

**Comment by dinakar29** [comment]
> Yeah, an orphan branch is created [here](https://github.com/insightsengineering/coverage-action/blob/179c7fd3ecaf44cbbfbe0aac188de22ec67b00d4/action.yml#L140). By empty do you mean that only the badge and the report should exist on this branch? Or can other files be there as well?

Alright, that's great. 

As for other files being there, presumably the only other files and directories that _should_ be there would be directories named after branches whose content is the XML and the SVG. A README.md at the top level is fine, but none of the original repository contents should be there for minimizing git storage.

File: action.yml [hunk 2]
```diff
@@ -68,26 +80,59 @@ runs:
...
+        # Switch to the branch if it already exists
+        git switch ${{ inputs.diff-storage }} || true
+        git pull origin ${{ inputs.diff-storage }} || true
+        # Create the branch if it doesn't exist yet
+        git checkout --orphan ${{ inputs.diff-storage }} || true
+        # Commands below will have no effect on the storage branch that already exists.
+        # No changes will be committed.
+        # Prepare the orphan branch containing only SVG badges and coverage report for diff-branch.
+        # Iterate through directories in top-level directory in repo.
+        # Remove anything, except .git, badges and coverage-report-action.
```
Comment (dinakar29) [inline action.yml:102]: Excellent documentation!

File: action.yml [hunk 3]
```diff
@@ -68,26 +80,59 @@ runs:
...
+        find . -mindepth 1 -maxdepth 1 ! \( -name .git -o -name badges -o -name coverage-report-action \) -type d -exec rm -rf {} +
+        # Remove other files in top-level directory.
+        find . -mindepth 1 -maxdepth 1 -type f -delete
+        # Initial commit has to contain some file
+        touch README.md
+        git add --all -f
+        git commit -m "Initialize storage branch" || true
```
Comment (dinakar29) [inline action.yml:109]: In the near future, we should make this its own Github Action. For now, this looks great.

File: action.yml [hunk 4]
```diff
@@ -98,10 +143,60 @@ runs:
...
+        github_token: ${{ inputs.token }}
+        branch: ${{ inputs.diff-storage }}
+        directory: ${{ inputs.diff-storage }}/badges
+
     - name: Generate issue comment body
       if: contains(inputs.publish, 'true')
       run: |
-        echo -e "## Code Coverage Summary\n" > .coverage-output.final
+        BRANCH=${GITHUB_HEAD_REF:-${GITHUB_REF#refs/heads/}}
+        echo -e "![badge](https://raw.githubusercontent.com/${{ github.repository_owner }}/${{ github.event.repository.name }}/${{ inputs.diff-storage }}/badges/$BRANCH/test-coverage.svg)\n\n" > .coverage-output.final
```
Comment (dinakar29) [inline action.yml:198]:
```suggestion
        echo -e "![badge](https://raw.githubusercontent.com/${{ github.repository_owner }}/${{ github.event.repository.name }}/${{ inputs.diff-storage }}/badges/${BRANCH}/test-coverage.svg)\n\n" > .coverage-output.final
```

File: action.yml [hunk 5]
```diff
@@ -126,6 +221,30 @@ runs:
         path: .coverage-output.final
         GITHUB_TOKEN: ${{ inputs.token }}
 
+    - name: Commit XML report to ${{ inputs.diff-storage }}
+      if: contains(inputs.diff, 'true') && startsWith(steps.branch-names.outputs.current_branch, inputs.diff-branch) && endsWith(steps.branch-names.outputs.current_branch, inputs.diff-branch)
```
Comment (dinakar29) [inline action.yml:225]:
```suggestion
      if: >
        contains(inputs.diff, 'true') &&
          startsWith(steps.branch-names.outputs.current_branch, inputs.diff-branch) &&
          endsWith(steps.branch-names.outputs.current_branch, inputs.diff-branch)
```

**Review by dinakar29 — COMMENTED** [review]
Looks great so far!

File: action.yml (hunk 2 above)
Comment (walkowif) [reply action.yml:102]: Thanks!

File: action.yml (hunk 3 above)
Comment (walkowif) [reply action.yml:109]: Okay.

**Comment by walkowif** [comment]
Example of PR with comparison of code coverage: https://github.com/walkowif/tern/pull/3
Contents of `_xml_coverage_reports` branch: https://github.com/walkowif/tern/tree/_xml_coverage_reports

**Comment by walkowif** [comment]
`|| github.event_name == 'push'` has to be added to `coverage` job, if one wants to generate update the coverage report on `main`.

For example:
```
coverage:
  if: github.event_name == 'pull_request' || github.event_name == 'push'
  name: Coverage 📔
  uses: insightsengineering/r.pkg.template/.github/workflows/test-coverage.yaml@main
  secrets:
    REPO_GITHUB_TOKEN: ${{ secrets.REPO_GITHUB_TOKEN }}
  with:
    publish-coverage-report: true
```

**Comment by cicdguy** [comment]
I noted is that the action doesn't store the XML coverage reports for the current branch in the `_xml_coverage_reports` branch.

So this is the directory structure that I was expecting:

```
🪵 _xml_coverage_reports 
❯ tree
.
├── README.md
├── branch_1
│   ├── badge.svg
│   └── coverage.xml
├── branch_2
│   ├── badge.svg
│   └── coverage.xml
├── branch_3
│   ├── badge.svg
│   └── coverage.xml
├── branch_N
│   ├── badge.svg
│   └── coverage.xml
├── main
│   ├── badge.svg
│   └── coverage.xml
├── pre-release
│   ├── badge.svg
│   └── coverage.xml
└── some-other-branch
    ├── badge.svg
    └── coverage.xml
```

Instead it looks like it's this:

```
🪵 _xml_coverage_reports 
❯ tree
.
├── README.md
├── badges
│   ├── branch_1
│   │   └── badge.svg
│   ├── branch_2
│   │   └── badge.svg
│   ├── branch_3
│   │   └── badge.svg
│   ├── branch_N
│   │   └── badge.svg
│   ├── main
│   │   └── badge.svg
│   ├── pre-release
│   │   └── badge.svg
│   └── some-other-branch
│       └── badge.svg
└── coverage-report-action
    └── coverage.xml
```

For provenance and posterity purposes, it's best to _also_ retain the coverage reports for the feature branches. This way if someone wants to manually do a diff across 2 branches, they will be able to do so.

**Comment by walkowif** [comment]
The [`_xml_coverage_reports`](https://github.com/walkowif/tern/tree/_xml_coverage_reports) branch now stores the reports for each feature branch. They have to be inside a directory with a known name (I chose `data`), because otherwise it will not be possible to clean the repo while initializing the storage branch.

File: action.yml [hunk 6]
```diff
@@ -126,6 +220,29 @@ runs:
...
+    - name: Commit XML report to ${{ inputs.diff-storage }}
+      if: >
+        contains(inputs.diff, 'true')
+      working-directory: ${{ inputs.diff-storage }}
+      run: |
+        git config --local user.email "action@github.com"
+        git config --local user.name "GitHub Actions"
+        git switch ${{ inputs.diff-storage }}
+        git pull origin ${{ inputs.diff-storage }}
+        mv ../coverage-report-action/coverage.xml ./data/${{ env.diff_storage_branch }}/coverage.xml
```
Comment (cicdguy) [inline action.yml:232]: Looks like this results in an error, per the tests on this repo.

Comment (walkowif) [reply action.yml]: Fixed, this was caused by the action not expecting coverage report to be in a differently named file and possibly in a subdirectory like here:
https://github.com/insightsengineering/coverage-action/blob/main/.github/workflows/test.yaml#L55

Two test attempts, both successful:
https://github.com/insightsengineering/coverage-action/actions/runs/2780956007/attempts/1
https://github.com/insightsengineering/coverage-action/actions/runs/2780956007

Comment (walkowif) [reply action.yml]: I also re-ran coverage action on the test repo:
https://github.com/walkowif/tern/pull/7
https://github.com/walkowif/tern/pull/8

**Pull request merged**